File: app/src/main/java/tpcreative/co/qrscanner/ui/history/HistoryFragment.java
package tpcreative.co.qrscanner.ui.history;
import android.Manifest;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.RelativeLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.content.ContextCompat;
import androidx.core.content.FileProvider;
import com.google.zxing.client.result.ParsedResultType;
import com.jaychang.srv.SimpleRecyclerView;
import com.jaychang.srv.decoration.SectionHeaderProvider;
import com.jaychang.srv.decoration.SimpleSectionHeaderProvider;
import com.karumi.dexter.Dexter;
import com.karumi.dexter.MultiplePermissionsReport;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.DexterError;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.PermissionRequestErrorListener;
import com.karumi.dexter.listener.multi.MultiplePermissionsListener;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import butterknife.BindView;
import de.mrapp.android.dialog.MaterialDialog;
import tpcreative.co.qrscanner.BuildConfig;
import tpcreative.co.qrscanner.R;
import tpcreative.co.qrscanner.common.BaseFragment;
import tpcreative.co.qrscanner.common.Navigator;
import tpcreative.co.qrscanner.common.HistorySingleton;
import tpcreative.co.qrscanner.common.MainSingleton;
import tpcreative.co.qrscanner.common.Utils;
import tpcreative.co.qrscanner.common.controller.ServiceManager;
import tpcreative.co.qrscanner.common.services.QRScannerApplication;
import tpcreative.co.qrscanner.helper.SQLiteHelper;
import tpcreative.co.qrscanner.model.Create;
import tpcreative.co.qrscanner.model.EnumFragmentType;
import tpcreative.co.qrscanner.model.HistoryModel;
import tpcreative.co.qrscanner.ui.scannerresult.ScannerResultFragment;

public class HistoryFragment extends BaseFragment implements HistoryView, HistoryCell.ItemSelectedListener, HistorySingleton.SingletonHistoryListener, MainSingleton.SingleTonMainListener {

    private static final String TAG = HistoryFragment.class.getSimpleName();
    @BindView(R.id.rlRoot)
    RelativeLayout rlRoot;
    @BindView(R.id.tvNotFoundItems)
    AppCompatTextView tvNotFoundItems;

    @BindView(R.id.recyclerView)
    SimpleRecyclerView recyclerView;
    private HistoryPresenter presenter;

    private boolean isDeleted;
    private boolean isSelectedAll = false;
    private ActionMode actionMode;


    private ActionMode.Callback callback = new ActionMode.Callback() {
        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            MenuInflater menuInflater = mode.getMenuInflater();
            menuInflater.inflate(R.menu.menu_select_all, menu);
            actionMode = mode;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                Window window = QRScannerApplication.getInstance().getActivity().getWindow();
                window.setStatusBarColor(ContextCompat.getColor(getContext(),R.color.colorAccentDark));
            }
            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            int i = item.getItemId();

            switch (item.getItemId()){
                case R.id.menu_item_select_all:{
                    final List<HistoryModel> list = presenter.getListGroup();
                    presenter.mList.clear();
                    if (isSelectedAll) {
                        for (HistoryModel index : list) {
                            index.setDeleted(true);
                            index.setChecked(false);
                            presenter.mList.add(index);
                        }
                        isSelectedAll = false;
                    } else {
                        for (HistoryModel index : list) {
                            index.setDeleted(true);
                            index.setChecked(true);
                            presenter.mList.add(index);
                        }
                        isSelectedAll = true;
                    }

                    if (actionMode!=null){
                        actionMode.setTitle(presenter.getCheckedCount() + " " + getString(R.string.selected));
                    }

                    recyclerView.removeAllCells();
                    bindData();
                    return true;
                }
                case R.id.menu_item_delete:{
                    final List<HistoryModel> listHistory = SQLiteHelper.getHistoryList();
                    if (listHistory==null){
                        return false;
                    }
                    if (listHistory.size()==0){
                        return false;
                    }
                    Log.d(TAG, "start");
                    if (presenter.getCheckedCount() > 0) {
                        dialogDelete();
                    }
                    return true;
                }
            }
            return false;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            actionMode = null;
            isSelectedAll = false;
            final List<HistoryModel> list = presenter.getListGroup();
            presenter.mList.clear();
            for (HistoryModel index : list) {
                index.setDeleted(false);
                presenter.mList.add(index);
            }
            recyclerView.removeAllCells();
            bindData();
            isDeleted = false;

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                Window window = QRScannerApplication.getInstance().getActivity().getWindow();
                window.setStatusBarColor(ContextCompat.getColor(getContext(),R.color.colorPrimaryDark));
            }
        }
    };

    public static HistoryFragment newInstance(int index) {
        HistoryFragment fragment = new HistoryFragment();
        Bundle b = new Bundle();
        b.putInt("index", index);
        fragment.setArguments(b);
        return fragment;
    }

    @Override
    protected int getLayoutId() {
        return 0;
    }

    @Override
    protected View getLayoutId(LayoutInflater inflater, ViewGroup viewGroup) {
        View view = inflater.inflate(R.layout.fragment_history, viewGroup, false);
        return view;
    }

    @Override
    protected void work() {
        super.work();
        HistorySingleton.getInstance().setListener(this);
        presenter = new HistoryPresenter();
        presenter.bindView(this);
        presenter.getListGroup();
        addRecyclerHeaders();
        bindData();
    }



    @Override
    public boolean isDeleted() {
        return isDeleted;
    }

    private void addRecyclerHeaders() {
        SectionHeaderProvider<HistoryModel> sh = new SimpleSectionHeaderProvider<HistoryModel>() {
            @NonNull
            @Override
            public View getSectionHeaderView(@NonNull HistoryModel history, int i) {
                View view = LayoutInflater.from(getContext()).inflate(R.layout.history_item_header, null, false);
                TextView textView = view.findViewById(R.id.tvHeader);
                textView.setText(history.getCategoryName());
                return view;
            }

            @Override
            public boolean isSameSection(@NonNull HistoryModel history, @NonNull HistoryModel nextHistory) {
                return history.getCategoryId() == nextHistory.getCategoryId();
            }

            // Optional, whether the header is sticky, default false
            @Override
            public boolean isSticky() {
                return false;
            }
        };
        recyclerView.setSectionHeader(sh);
    }

    private void bindData() {
        List<HistoryModel> mListItems = presenter.mList;
        List<HistoryCell> cells = new ArrayList<>();
        //LOOP THROUGH GALAXIES INSTANTIATING THEIR CELLS AND ADDING TO CELLS COLLECTION
        for (HistoryModel items : mListItems) {
            HistoryCell cell = new HistoryCell(items);
            cell.setListener(this);
            cells.add(cell);
        }
        if (mListItems!=null){
            if (mListItems.size()>0){
                tvNotFoundItems.setVisibility(View.INVISIBLE);
            }
            else {
                tvNotFoundItems.setVisibility(View.VISIBLE);
            }
        }
        recyclerView.addCells(cells);
    }

    @Override
    public Context getContext() {
        return getActivity();
    }

    @Override
    public void onClickItem(int position, boolean isChecked) {
        Log.d(TAG, "position : " + position + " - " + isChecked);
        boolean result = presenter.mList.get(position).isDeleted();
        presenter.mList.get(position).setChecked(isChecked);
        if (result) {
            if (actionMode!=null){
                actionMode.setTitle(presenter.getCheckedCount() + " " + getString(R.string.selected));
            }
        }
    }


    @Override
    public void isShowDeleteAction(boolean isDelete) {
        final List<HistoryModel> listHistory = SQLiteHelper.getHistoryList();
        if (isDelete) {
            if (actionMode == null) {
                actionMode = QRScannerApplication.getInstance().getActivity().getToolbar().startActionMode(callback);
            }
            if (listHistory == null) {

                return;

            }
            if (listHistory.size() == 0) {

                return;
            }

            final List<HistoryModel> list = presenter.getListGroup();
            presenter.mList.clear();
            for (HistoryModel index : list) {
                index.setDeleted(true);
                presenter.mList.add(index);
            }
            recyclerView.removeAllCells();
            bindData();
            isDeleted = true;
        } else {
            if (listHistory == null) {
                return;
            }
            if (listHistory.size() == 0) {
                return;
            }
            onAddPermissionSave();
        }
    }


    @Override
    public void onLongClickItem(int position){
        if (actionMode == null) {
            actionMode = QRScannerApplication.getInstance().getActivity().getToolbar().startActionMode(callback);
        }
        final List<HistoryModel> list = presenter.getListGroup();
        presenter.mList.clear();
        for (HistoryModel index : list) {
            index.setDeleted(true);
            presenter.mList.add(index);
        }
        presenter.mList.get(position).setChecked(true);
        if (actionMode!=null){
            actionMode.setTitle(presenter.getCheckedCount() + " " + getString(R.string.selected));
        }
        recyclerView.removeAllCells();
        bindData();
        isDeleted = true;
    }

    @Override
    public void onClickItem(int position) {
        if (actionMode!=null){
            return;
        }
        final Create create = new Create();
        final HistoryModel history = presenter.mList.get(position);
        if (history.createType.equalsIgnoreCase(ParsedResultType.ADDRESSBOOK.name())) {
            create.address = history.address;
            create.fullName = history.fullName;
            create.email = history.email;
            create.phone = history.phone;
            create.createType = ParsedResultType.ADDRESSBOOK;
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.EMAIL_ADDRESS.name())) {
            create.email = history.email;
            create.subject = history.subject;
            create.message = history.message;
            create.createType = ParsedResultType.EMAIL_ADDRESS;
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.PRODUCT.name())) {
            create.createType = ParsedResultType.PRODUCT;
            create.productId = history.text;
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.URI.name())) {
            create.url = history.url;
            create.createType = ParsedResultType.URI;
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.WIFI.name())) {
            create.hidden = history.hidden;
            create.ssId = history.ssId;
            create.networkEncryption = history.networkEncryption;
            create.password = history.password;
            create.createType = ParsedResultType.WIFI;
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.GEO.name())) {
            create.lat = history.lat;
            create.lon = history.lon;
            create.query = history.query;
            create.createType = ParsedResultType.GEO;
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.TEL.name())) {
            create.phone = history.phone;
            create.createType = ParsedResultType.TEL;
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.SMS.name())) {
            create.phone = history.phone;
            create.message = history.message;
            create.createType = ParsedResultType.SMS;
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.CALENDAR.name())) {
            create.title = history.title;
            create.description = history.description;
            create.location = history.location;
            create.startEvent = history.startEvent;
            create.endEvent = history.endEvent;
            create.startEventMilliseconds = history.startEventMilliseconds;
            create.endEventMilliseconds = history.endEventMilliseconds;
            create.createType = ParsedResultType.CALENDAR;
        }
        else if (history.createType.equalsIgnoreCase(ParsedResultType.ISBN.name())){
            create.ISBN = history.text;
            create.createType = ParsedResultType.ISBN;
        }
        else {
            create.text = history.text;
            create.createType = ParsedResultType.TEXT;
        }
        create.fragmentType = EnumFragmentType.HISTORY;
        Navigator.onResultView(getActivity(),create,ScannerResultFragment.class);
    }

    @Override
    public void onClickShare(int position) {
        final HistoryModel history = presenter.mList.get(position);
        StringBuilder sb = new StringBuilder();
        if (history.createType.equalsIgnoreCase(ParsedResultType.ADDRESSBOOK.name())) {
            sb.append("Address :"+history.address);
            sb.append("\n");
            sb.append("FullName :"+ history.fullName);
            sb.append("\n");
            sb.append("Email :"+ history.email);
            sb.append("\n");
            sb.append("Phone :"+history.phone);
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.EMAIL_ADDRESS.name())) {
            sb.append("Email :"+history.email);
            sb.append("\n");
            sb.append("Subject :"+ history.subject);
            sb.append("\n");
            sb.append("Message :"+ history.message);
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.PRODUCT.name())) {
            sb.append("ProductId :"+history.text);
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.URI.name())) {
            sb.append("Url :"+history.url);
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.WIFI.name())) {
            sb.append("SSId :"+history.ssId);
            sb.append("\n");
            sb.append("Password :"+ history.password);
            sb.append("\n");
            sb.append("Network encryption :"+ history.networkEncryption);
            sb.append("\n");
            sb.append("Hidden :"+ history.hidden);
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.GEO.name())) {
            sb.append("Latitude :"+history.lat);
            sb.append("\n");
            sb.append("Longitude :"+ history.lon);
            sb.append("\n");
            sb.append("Query :"+ history.query);

        } else if (history.createType.equalsIgnoreCase(ParsedResultType.TEL.name())) {
            sb.append("Phone :"+history.phone);
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.SMS.name())) {
            sb.append("Phone :"+history.phone);
            sb.append("\n");
            sb.append("Message :"+ history.message);
        } else if (history.createType.equalsIgnoreCase(ParsedResultType.CALENDAR.name())) {
            sb.append("Title :"+history.title);
            sb.append("\n");
            sb.append("Description :"+ history.description);
            sb.append("\n");
            sb.append("Location :"+ history.location);
            sb.append("\n");
            sb.append("Start event :"+history.startEvent);
            sb.append("\n");
            sb.append("End event :"+ history.endEvent);
        }
        else if (history.createType.equalsIgnoreCase(ParsedResultType.ISBN.name())){
            sb.append("ISBN :"+history.text);
        }
        else {
            sb.append("Text :"+history.text);
        }
        shareToSocial(sb.toString());
    }

    public void shareToSocial(String value){
        Intent intent = new Intent();
        intent.setAction(Intent.ACTION_SEND);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_TEXT,value);
        startActivity(Intent.createChooser(intent, "Share"));
    }

    public void shareToSocial(final Uri value) {
        Log.d(TAG, "path call");
        Intent intent = new Intent();
        intent.setAction(Intent.ACTION_SEND);
        intent.setType("image/*");
        intent.putExtra(Intent.EXTRA_STREAM, value);
        startActivity(Intent.createChooser(intent, "Share"));
    }

    @Override
    public void reloadData() {
        if (presenter !=null){
            if(recyclerView!=null){
                presenter.getListGroup();
                recyclerView.removeAllCells();
                bindData();
            }
        }
    }

    public void onAddPermissionSave() {
        Dexter.withContext(getActivity())
                .withPermissions(
                        Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE)
                .withListener(new MultiplePermissionsListener() {
                    @Override
                    public void onPermissionsChecked(MultiplePermissionsReport report) {
                        if (report.areAllPermissionsGranted()) {
                            ServiceManager.getInstance().onExportDatabaseCSVTask(EnumFragmentType.HISTORY, new ServiceManager.ServiceManagerListener() {
                                @Override
                                public void onExportingSVCCompleted(String path) {
                                    File file = new File(path);
                                    if (file.isFile()) {
                                        Log.d(TAG, "path : " + path);
                                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                                            Uri uri = FileProvider.getUriForFile(getContext(), BuildConfig.APPLICATION_ID + ".provider", file);
                                            shareToSocial(uri);
                                        } else {
                                            Uri uri = Uri.fromFile(file);
                                            shareToSocial(uri);
                                        }
                                    }
                                }
                            });

                        } else {
                            Log.d(TAG, "Permission is denied");
                        }
                        // check for permanent denial of any permission
                        if (report.isAnyPermissionPermanentlyDenied()) {
                            /*Miss add permission in manifest*/
                            Log.d(TAG, "request permission is failed");
                        }
                    }

                    @Override
                    public void onPermissionRationaleShouldBeShown(List<PermissionRequest> permissions, PermissionToken token) {
                        /* ... */
                        token.continuePermissionRequest();
                    }
                })
                .withErrorListener(new PermissionRequestErrorListener() {
                    @Override
                    public void onError(DexterError error) {
                        Log.d(TAG, "error ask permission");
                    }
                }).onSameThread().check();
    }

    @Override
    public void setMenuVisibility(boolean menuVisible) {
        super.setMenuVisibility(menuVisible);
        if (menuVisible) {
            Log.d(TAG,"isVisible");
            MainSingleton.getInstance().setListener(this);
            QRScannerApplication.getInstance().getActivity().onShowFloatingButton(HistoryFragment.this,true);
        }
        else{
            MainSingleton.getInstance().setListener(null);
            Log.d(TAG,"isInVisible");
            if (actionMode!=null){
                actionMode.finish();
            }
        }
    }

    public void dialogDelete() {
        MaterialDialog.Builder builder = new MaterialDialog.Builder(getContext(), Utils.getCurrentTheme());
        builder.setTitle(getString(R.string.delete));
        builder.setMessage(String.format(getString(R.string.dialog_delete),presenter.getCheckedCount()+""));
        builder.setNegativeButton(getString(R.string.no), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {

            }
        });
        builder.setPositiveButton(getString(R.string.yes), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                presenter.deleteItem();
                isSelectedAll = false;
                isDeleted = false;
                if (actionMode!=null){
                    actionMode.finish();
                }
            }
        });
        builder.show();
    }

    @Override
    public void updateView() {
        recyclerView.removeAllCells();
        bindData();
    }

    @Override
    public void onStop() {
        super.onStop();
        Log.d(TAG, "onStop");
    }

    @Override
    public void onStart() {
        super.onStart();
        Log.d(TAG, "onStart");
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.d(TAG, "onDestroy");
    }

    @Override
    public void onResume() {
        super.onResume();
        Log.d(TAG, "onResume");
    }
}

File: app/src/main/java/tpcreative/co/qrscanner/ui/history/HistoryFragment.kt
package tpcreative.co.qrscanner.ui.history
import android.app.Activity
import android.app.Dialog
import android.content.ClipData
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.*
import android.widget.TextView
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import co.tpcreative.supersafe.common.network.Status
import com.afollestad.materialdialogs.MaterialDialog
import com.jaychang.srv.decoration.SectionHeaderProvider
import com.jaychang.srv.decoration.SimpleSectionHeaderProvider
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import tpcreative.co.qrscanner.BuildConfig
import tpcreative.co.qrscanner.R
import tpcreative.co.qrscanner.common.*
import tpcreative.co.qrscanner.common.controller.ServiceManager
import tpcreative.co.qrscanner.common.extension.toText
import tpcreative.co.qrscanner.common.services.QRScannerApplication
import tpcreative.co.qrscanner.databinding.FragmentHistoryBinding
import tpcreative.co.qrscanner.helper.SQLiteHelper
import tpcreative.co.qrscanner.model.*
import tpcreative.co.qrscanner.ui.scannerresult.ScannerResultActivity
import java.io.File
import java.util.*

class HistoryFragment : BaseFragment(), HistoryCell.ItemSelectedListener, HistorySingleton.SingletonHistoryListener{
    lateinit var viewModel : HistoryViewModel
    var misDeleted = false
    var isSelectedAll = false
    var actionMode: ActionMode? = null
    var dialog : Dialog? = null
    var dialogExport : Dialog? = null
    lateinit var binding : FragmentHistoryBinding
    private val callback: ActionMode.Callback = object : ActionMode.Callback {
        override fun onCreateActionMode(mode: ActionMode?, menu: Menu?): Boolean {
            val menuInflater: MenuInflater? = mode?.menuInflater
            menuInflater?.inflate(R.menu.menu_select_all, menu)
            actionMode = mode
            val window: Window? = QRScannerApplication.getInstance().getActivity()?.window
            window?.statusBarColor = ContextCompat.getColor(context!!, R.color.colorAccentDark)
            binding.rlDelete.visibility = View.INVISIBLE
            binding.rlCSV.visibility = View.INVISIBLE
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode?, menu: Menu?): Boolean {
            return false
        }

        override fun onActionItemClicked(mode: ActionMode?, item: MenuItem?): Boolean {
            when (item?.itemId) {
                R.id.menu_item_select_all -> {
                    val list: MutableList<HistoryModel> = viewModel.getListGroup()
                    viewModel.mList.clear()
                    isSelectedAll = if (isSelectedAll) {
                        for (index in list) {
                            index.setDeleted(true)
                            index.setChecked(false)
                            viewModel.mList.add(index)
                        }
                        false
                    } else {
                        for (index in list) {
                            index.setDeleted(true)
                            index.setChecked(true)
                            viewModel.mList.add(index)
                        }
                        true
                    }
                    if (actionMode != null) {
                        actionMode?.title = viewModel.getCheckedCount().toString() + " " + getString(R.string.selected)
                    }
                    binding.recyclerView.removeAllCells()
                    bindData()
                    return true
                }
                R.id.menu_item_delete -> {
                    val listHistory: MutableList<HistoryModel> = SQLiteHelper.getHistoryList()
                    if (listHistory.size == 0) {
                        return false
                    }
                    Utils.Log(TAG, "start")
                    if (viewModel.getCheckedCount() > 0) {
                        dialog?.show()
                        deleteItem()
                    }
                    return true
                }
            }
            return false
        }

        override fun onDestroyActionMode(mode: ActionMode?) {
            actionMode = null
            isSelectedAll = false
            binding.rlDelete.visibility = View.VISIBLE
            binding.rlCSV.visibility = View.VISIBLE
            val list: MutableList<HistoryModel> = viewModel.getListGroup()
            viewModel.mList.clear()
            for (index in list) {
                index.setDeleted(false)
                viewModel.mList.add(index)
            }
            binding.recyclerView.removeAllCells()
            bindData()
            misDeleted = false
            val window: Window? = QRScannerApplication.getInstance().getActivity()?.window
            window?.statusBarColor = ContextCompat.getColor(context!!, R.color.colorPrimaryDark)
        }
    }

    override fun getLayoutId(): Int {
        return 0
    }

    override fun getLayoutId(inflater: LayoutInflater?, viewGroup: ViewGroup?): View? {
        binding = FragmentHistoryBinding.inflate(layoutInflater)
        return binding.root
    }

    override fun work() {
        super.work()
        initUI()
        dialog = ProgressDialog.progressDialog(requireContext(),R.string.waiting_for_delete.toText())
        dialogExport = ProgressDialog.progressDialog(requireContext(),R.string.waiting_for_export.toText())
    }

    override fun isDeleted(): Boolean {
        return misDeleted
    }

    fun addRecyclerHeaders() {
        val sh: SectionHeaderProvider<HistoryModel?> = object : SimpleSectionHeaderProvider<HistoryModel?>() {
            override fun getSectionHeaderView(history: HistoryModel, i: Int): View {
                val view: View = LayoutInflater.from(context).inflate(R.layout.history_item_header, null, false)
                val textView: TextView = view.findViewById<TextView?>(R.id.tvHeader)
                textView.text = history.getCategoryName()
                return view
            }

            override fun isSameSection(history: HistoryModel, nextHistory: HistoryModel): Boolean {
                return history.getCategoryId() == nextHistory.getCategoryId()
            }

            // Optional, whether the header is sticky, default false
            override fun isSticky(): Boolean {
                return false
            }
        }
        binding.recyclerView.setSectionHeader(sh)
    }

    fun bindData() {
        val mListItems: MutableList<HistoryModel> = viewModel.mList
        val cells: MutableList<HistoryCell?> = ArrayList()
        //LOOP THROUGH GALAXIES INSTANTIATING THEIR CELLS AND ADDING TO CELLS COLLECTION
        for (items in mListItems) {
            val cell = HistoryCell(items)
            cell.setListener(this)
            cells.add(cell)
        }
        if (mListItems.size > 0) {
            binding.tvNotFoundItems.visibility = View.INVISIBLE
        } else {
            binding.tvNotFoundItems.visibility = View.VISIBLE
        }
        binding.recyclerView.addCells(cells)
        binding.tvCount.text = "${viewModel.count()}"
    }

    override fun getContext(): Context? {
        return activity
    }

    override fun onClickItem(position: Int, isChecked: Boolean) {
        Utils.Log(TAG, "position : $position - $isChecked")
        val result = viewModel.mList[position].isDeleted()
        viewModel.mList[position].setChecked(isChecked)
        Utils.Log(TAG,"isChecked $result")
        Utils.Log(TAG,"isDeleted ${viewModel.mList[position].isChecked()}")
        if (result) {
            if (actionMode != null) {
                actionMode?.title = viewModel.getCheckedCount().toString() + " " + getString(R.string.selected)
            }
        }
    }

    override fun onLongClickItem(position: Int) {
        if (actionMode == null) {
            actionMode = QRScannerApplication.getInstance().getActivity()?.getToolbar()?.startActionMode(callback)
        }
        val list: MutableList<HistoryModel> = viewModel.getListGroup()
        viewModel.mList.clear()
        for (index in list) {
            index.setDeleted(true)
            viewModel.mList.add(index)
        }
        viewModel.mList[position].setChecked(true)
        if (actionMode != null) {
            actionMode?.title = viewModel.getCheckedCount().toString() + " " + getString(R.string.selected)
        }
        binding.recyclerView.removeAllCells()
        bindData()
        misDeleted = true
    }

    override fun onClickItem(position: Int) {
        if (actionMode != null) {
            return
        }
        val history: HistoryModel = viewModel.mList[position]
        val create = GeneralModel(history,EnumFragmentType.HISTORY,EnumImplement.VIEW)
        viewForResult.launch(Navigator.onResultView(activity, create, ScannerResultActivity::class.java))
    }

    private val viewForResult = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK) {
            Utils.Log(TAG,"${result.resultCode}")
        }
    }

    private fun shareToSocial(value: Uri?) {
        Utils.Log(TAG, "path call")
        val intent = Intent()
        intent.action = Intent.ACTION_SEND
        intent.type = "*/*"
        intent.putExtra(Intent.EXTRA_STREAM, value)
        intent.clipData = ClipData.newRawUri("", value);
        intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION or Intent.FLAG_GRANT_READ_URI_PERMISSION)
        startActivity(Intent.createChooser(intent, "Share"))
    }

    override fun reloadData() {
        CoroutineScope(Dispatchers.Main).launch {
            if (binding.recyclerView != null) {
                viewModel.getListGroup()
                binding.recyclerView.removeAllCells()
                bindData()
            }
        }
    }

    fun exportData() = CoroutineScope(Dispatchers.Main).launch {
        val mResult =  ServiceManager.getInstance().onExportDatabaseCSVTask(requireContext(),EnumFragmentType.HISTORY)
        when(mResult.status){
            Status.SUCCESS -> {
                val path = mResult.data ?: ""
                val file = File(path)
                if (file.isFile) {
                    Utils.Log(TAG, "path : $path")
                    val uri: Uri = FileProvider.getUriForFile(requireContext(), BuildConfig.APPLICATION_ID + ".provider", file)
                    shareToSocial(uri)
                }
                dialogExport?.dismiss()
            }else -> {
            Utils.Log(TAG,"")
            dialogExport?.dismiss()
        }
        }
    }

    override fun setMenuVisibility(menuVisible: Boolean) {
        super.setMenuVisibility(menuVisible)
        if (menuVisible) {
            if (actionMode != null) {
                actionMode?.finish()
            }
        }
    }

    fun dialogEntireDelete() {
        MaterialDialog(requireContext()).show {
            title(R.string.delete)
            message(text = getString(R.string.delete_entire_history))
            positiveButton(R.string.yes){
                dialog?.show()
                deleteEntireItem()
            }
            negativeButton (R.string.no){
            }
        }
    }

    fun updateView() {
        binding.recyclerView.removeAllCells()
        bindData()
    }

    override fun onStop() {
        super.onStop()
        Utils.Log(TAG, "onStop")
    }

    override fun onStart() {
        super.onStart()
        Utils.Log(TAG, "onStart")
    }

    override fun onDestroy() {
        super.onDestroy()
        Utils.Log(TAG, "onDestroy")
    }

    override fun onResume() {
        super.onResume()
        Utils.Log(TAG, "onResume")
    }

    companion object {
        private val TAG = HistoryFragment::class.java.simpleName
        fun newInstance(index: Int): HistoryFragment {
            val fragment = HistoryFragment()
            val b = Bundle()
            b.putInt("index", index)
            fragment.arguments = b
            return fragment
        }
    }
}